### DISCUSSION TITLE
Label entire substructure match with recursive SMARTS

### DISCUSSION THREAD

**Discussion by Versallius** [rdkit/rdkit#6484] [Q&A]
Hi, 

I am trying to find which atoms in a molecule match a fragment from a SMARTS string. Currently GetSubstructMatches works for normal SMARTS formats but not recursive smarts. In the example below, I want the match to return ((9, 10, 11, 12, 13, 14, 15, 16),)​, since these are the atoms that are part of the heterocyclic nitro substructure encoded within the SMARTS, instead​ it only returned atom 12.

```
testsmarts = '[$(c[N+](=O)[O-]);r5]'
testmol = Chem.MolFromSmiles('CCC(C)OC(=O)C=Cc1ccc([N+](=O)[O-])o1')
testfrag = Chem.MolFromSmarts(testsmarts)
for atom in testmol.GetAtoms():
    atom.SetProp("atomNote", str(atom.GetIdx()))

testmol.GetSubstructMatches(testfrag)
```

![image](https://github.com/rdkit/rdkit/assets/71128748/6cd05dd0-f7ff-4c98-b05b-6eaee3df61d4)

**Comment by DavidACosgrove** [comment]
When using recursive SMARTS, only the first (head) atom is returned in the
match.  This has been the case since Daylight defined the language decades
ago, it is not an RDKit-specific thing.

Also, in future please don't paste an image of your code ;paste it
separately, protected by triple back-ticks.  That way it is easier for
someone to reproduce your issue in their own environment by cutting and
pasting it. Otherwise they will have to type it in again, which raises the
activation energy and makes it less likely for you to get a reply.



On Thu, Jun 22, 2023 at 4:41 PM Charles Gong ***@***.***>
wrote:

> Hi,
>
> I am trying to find which atoms in a molecule match a fragment from a
> SMARTS string. Currently GetSubstructMatches works for normal SMARTS
> formats but not recursive smarts. In the example below, I want the match to
> return ((9, 10, 11, 12, 13, 14, 15, 16),)​, since these are the atoms that
> are part of the heterocyclic nitro substructure encoded within the SMARTS,
> instead​ it only returned atom 12.
>
> [image: image]
> <https://user-images.githubusercontent.com/71128748/247983251-6cd05dd0-f7ff-4c98-b05b-6eaee3df61d4.png>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/rdkit/rdkit/discussions/6484>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACGF2FR5EYO7HDE4YPTB3YTXMRRSDANCNFSM6AAAAAAZQMWBYQ>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 
David Cosgrove
Freelance computational chemistry and chemoinformatics developer
http://cozchemix.co.uk

**Reply by Versallius** [reply]
Thanks for the response! I understand that this is how recursive SMARTS works, but I was wondering if there was any way to also match the other atoms.

**Comment by DavidACosgrove** [comment]
In that case, you will have to specify all the other atoms, something like
[O-][N+](=O)c1****1 which I haven't tested but looks right.

On Thu, Jun 22, 2023 at 5:20 PM Charles Gong ***@***.***>
wrote:

> Thanks for the response! I understand that this is how recursive SMARTS
> works, but I was wondering if there was any way to also match the other
> atoms.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/rdkit/rdkit/discussions/6484#discussioncomment-6254027>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACGF2FUJRDPG2PVW7AWAIJLXMRWEHANCNFSM6AAAAAAZQMWBYQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
David Cosgrove
Freelance computational chemistry and chemoinformatics developer
http://cozchemix.co.uk

**Reply by Versallius** [reply]
I see, thank you. The problem is that I have 65 these recursive SMARTS to match and some of them are much more complicated than this example. I would like to avoid manually writing non-recursive SMARTS if possible (although this is a solution of last resort). Do you know of any way to algorithmically convert recursive SMARTS to non-recursive SMARTS?

**Comment by DavidACosgrove** [comment]
I don't think that that would be possible in the general case.  How would
you handle [$(c[N+]([O-]=O);r5,r6], for example?

On Thu, Jun 22, 2023 at 5:39 PM Charles Gong ***@***.***>
wrote:

> I see, thank you. The problem is that I have 65 these recursive SMARTS to
> match and some of them are much more complicated than this example. I would
> like to avoid manually writing non-recursive SMARTS if possible (although
> this is a solution of last resort). Do you know of any way to
> algorithmically convert recursive SMARTS to non-recursive SMARTS?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/rdkit/rdkit/discussions/6484#discussioncomment-6254186>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACGF2FTOZ2RYCGYGCBXUH4DXMRYKVANCNFSM6AAAAAAZQMWBYQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
David Cosgrove
Freelance computational chemistry and chemoinformatics developer
http://cozchemix.co.uk
